Fix `pyproject.toml` and stop using `setup.py`

IvanIsCoding opened the pull request:
<!--
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I ran rustfmt locally
- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

Related to #1492 

I tested this locally and it worked e.g. `python setup.py sdist`  passes.

For `retworkx`, it had more to do with the `dist` being in the wrong path

mtreinish reviewed (commented):
We probably should move the sdist creation to using `python -m build --sdist` for rustworkx. 
 When I test this locally it's able to build the sdist without errors like the CI job had.
 
The build isolation from `build` probably won't work for retworkx because that's a bit more special but that's ok.

IvanIsCoding commented on .github/workflows/wheels.yml:
```diff
@@ -2,28 +2,15 @@
   push:
-    tags:
-      - '*'
+    branches:
+      - 'stable/0.17'
```
I added an `if` condition to check that the date is today, does that work?

IvanIsCoding commented:
@mtreinish  I updated rustworkx's command to be `python -m build --sdist`.

For the one-off branch: I don't want to create yet another branch. Maybe we can reuse `stable/0.17`? I just added a check for today's date. That way we cannot trigger it by accident on the future.

mtreinish commented on .github/workflows/wheels.yml:
```diff
@@ -205,7 +47,7 @@ jobs:
       - uses: actions/upload-artifact@v4
         with:
-          path: ./retworkx/dist/*
+          path: ./dist/*
```
This isn't actually correct. The wheel gets generated in `./retworkx/dist`

mtreinish reviewed (commented):
Overall this is about ready, but in the interest of getting the fix out I pushed up a quick throwaway branch with this branch to manually trigger a build. The fixes are live now on pypi: https://pypi.org/project/retworkx/ and https://pypi.org/project/rustworkx/0.17.1/#files

Do you want to retarget this PR to main and then we can backport it to `stable/0.17`?

IvanIsCoding commented:
@mergifyio backport stable/0.17
